File: gneva/bot/audio_capture.js
/**
 * Gneva Audio Capture — injected into meeting pages via CDP.
 * Captures per-speaker audio from WebRTC tracks, sends PCM over WebSocket
 * with track ID prefix for speaker separation.
 *
 * Each remote audio track gets its own AudioContext + ScriptProcessorNode.
 * Audio chunks are sent as: [4-byte track ID (uint32 BE)] + [PCM16 data]
 */
(function() {
  const WS_URL = '__WS_URL__';  // replaced by Python before injection
  const SAMPLE_RATE = 16000;
  const BUFFER_SIZE = 4096;

  let ws = null;
  let reconnectTimer = null;

  // Per-track processing chains: trackKey -> { ctx, source, processor, trackId }
  const trackChains = new Map();
  let nextTrackId = 1;

  // Track IDs assigned by track key (stream.id or track.id)
  const trackIdMap = new Map();

  // Connected media elements (for DOM-based capture fallback)
  const connectedElements = new Set();

  // Shared AudioContext for DOM element capture (fallback path)
  let fallbackCtx = null;
  let fallbackProcessor = null;
  const FALLBACK_TRACK_ID = 0;  // track ID 0 = mixed/unknown source

  function connectWS() {
    if (ws && ws.readyState === WebSocket.OPEN) return;
    try {
      ws = new WebSocket(WS_URL);
      ws.binaryType = 'arraybuffer';
      ws.onopen = () => console.log('[Gneva] Audio capture WebSocket connected');
      ws.onclose = () => {
        console.log('[Gneva] WebSocket closed, reconnecting in 2s...');
        if (!reconnectTimer) reconnectTimer = setTimeout(() => { reconnectTimer = null; connectWS(); }, 2000);
      };
      ws.onerror = (e) => console.error('[Gneva] WebSocket error:', e);
    } catch (e) {
      console.error('[Gneva] WebSocket connect failed:', e);
    }
  }

  function getTrackId(key) {
    if (trackIdMap.has(key)) return trackIdMap.get(key);
    const id = nextTrackId++;
    trackIdMap.set(key, id);
    return id;
  }

  /**
   * Send PCM audio with a 4-byte track ID header.
   * Format: [uint32 BE trackId][int16 PCM samples...]
   */
  function sendWithTrackId(trackId, float32Input) {
    if (!ws || ws.readyState !== WebSocket.OPEN) return;
    // 4 bytes header + 2 bytes per sample
    const buf = new ArrayBuffer(4 + float32Input.length * 2);
    const view = new DataView(buf);
    // Write track ID as big-endian uint32
    view.setUint32(0, trackId, false);
    // Convert float32 to int16
    const pcm16 = new Int16Array(buf, 4);
    for (let i = 0; i < float32Input.length; i++) {
      const s = Math.max(-1, Math.min(1, float32Input[i]));
      pcm16[i] = s < 0 ? s * 0x8000 : s * 0x7FFF;
    }
    ws.send(buf);
  }

  /**
   * Create a per-track audio processing chain.
   * Each track gets its own AudioContext → MediaStreamSource → ScriptProcessor → WS.
   */
  function createTrackChain(stream, trackKey) {
    if (trackChains.has(trackKey)) {
      console.log('[Gneva] Track chain already exists for:', trackKey);
      return;
    }

    const trackId = getTrackId(trackKey);

    try {
      const ctx = new (window.AudioContext || window.webkitAudioContext)({ sampleRate: SAMPLE_RATE });
      const source = ctx.createMediaStreamSource(stream);
      const processor = ctx.createScriptProcessor(BUFFER_SIZE, 1, 1);

      processor.onaudioprocess = function(e) {
        const input = e.inputBuffer.getChannelData(0);
        sendWithTrackId(trackId, input);
      };

      source.connect(processor);
      // Connect to destination to keep the processor alive (required by spec)
      processor.connect(ctx.destination);

      const chain = { ctx, source, processor, trackId, trackKey };
      trackChains.set(trackKey, chain);

      // Listen for track ended to clean up
      const audioTracks = stream.getAudioTracks();
      if (audioTracks.length > 0) {
        audioTracks.forEach(track => {
          track.addEventListener('ended', () => {
            console.log('[Gneva] Audio track ended:', trackKey, '(trackId:', trackId, ')');
            cleanupTrackChain(trackKey);
          });
        });
      }

      console.log('[Gneva] Created per-track chain: trackKey=' + trackKey + ', trackId=' + trackId +
                   ', audioTracks=' + audioTracks.length);
    } catch (e) {
      console.warn('[Gneva] Failed to create track chain for', trackKey, ':', e.message);
    }
  }

  function cleanupTrackChain(trackKey) {
    const chain = trackChains.get(trackKey);
    if (!chain) return;
    try {
      chain.processor.disconnect();
      chain.source.disconnect();
      chain.ctx.close();
    } catch (e) {
      console.warn('[Gneva] Cleanup error for', trackKey, ':', e.message);
    }
    trackChains.delete(trackKey);
    console.log('[Gneva] Cleaned up track chain:', trackKey, '(active chains:', trackChains.size, ')');
  }

  /**
   * Initialize fallback AudioContext for DOM media element capture.
   * Used when audio comes through <audio>/<video> elements instead of direct WebRTC tracks.
   * This is the mixed-audio path (track ID 0) — less ideal but works as fallback.
   */
  function initFallbackAudio() {
    if (fallbackCtx) return;
    fallbackCtx = new (window.AudioContext || window.webkitAudioContext)({ sampleRate: SAMPLE_RATE });

    fallbackProcessor = fallbackCtx.createScriptProcessor(BUFFER_SIZE, 1, 1);
    fallbackProcessor.onaudioprocess = function(e) {
      const input = e.inputBuffer.getChannelData(0);
      sendWithTrackId(FALLBACK_TRACK_ID, input);
    };

    // We'll connect sources to a merger, then merger to processor
    window.__gnevaFallbackMerger = fallbackCtx.createChannelMerger(1);
    window.__gnevaFallbackMerger.connect(fallbackProcessor);
    fallbackProcessor.connect(fallbackCtx.destination);

    console.log('[Gneva] Fallback audio context initialized (track ID 0 = mixed)');
  }

  function captureElement(el) {
    if (connectedElements.has(el)) return;
    if (!el.srcObject && !el.src) return;

    try {
      initFallbackAudio();
      let source;

      if (el.srcObject) {
        source = fallbackCtx.createMediaStreamSource(el.srcObject);
      } else if (el.src && el.src.startsWith('blob:')) {
        source = fallbackCtx.createMediaElementSource(el);
      } else {
        return;
      }

      source.connect(window.__gnevaFallbackMerger);
      connectedElements.add(el);
      console.log('[Gneva] Capturing audio element (fallback path):', el.tagName, el.id || '(no id)');
    } catch (e) {
      if (!e.message.includes('already connected')) {
        console.warn('[Gneva] Could not capture element:', e.message);
      }
    }
  }

  function scanForMedia() {
    document.querySelectorAll('audio, video').forEach(el => {
      captureElement(el);

      if (!connectedElements.has(el)) {
        const origSet = Object.getOwnPropertyDescriptor(HTMLMediaElement.prototype, 'srcObject')?.set;
        if (origSet && !el.__gneva_hooked) {
          el.__gneva_hooked = true;
          Object.defineProperty(el, 'srcObject', {
            set(v) {
              origSet.call(this, v);
              setTimeout(() => captureElement(this), 100);
            },
            get() {
              return Object.getOwnPropertyDescriptor(HTMLMediaElement.prototype, 'srcObject').get.call(this);
            }
          });
        }
      }
    });
  }

  // Observe DOM for dynamically added media elements
  const observer = new MutationObserver((mutations) => {
    for (const m of mutations) {
      for (const node of m.addedNodes) {
        if (node.nodeType !== 1) continue;
        if (node.tagName === 'AUDIO' || node.tagName === 'VIDEO') {
          setTimeout(() => captureElement(node), 200);
        }
        if (node.querySelectorAll) {
          node.querySelectorAll('audio, video').forEach(el => {
            setTimeout(() => captureElement(el), 200);
          });
        }
      }
    }
  });
  observer.observe(document.body || document.documentElement, { childList: true, subtree: true });

  // Hook RTCPeerConnection to catch WebRTC audio tracks — per-track separation
  const OrigRTC = window.RTCPeerConnection;
  if (OrigRTC) {
    window.RTCPeerConnection = function(...args) {
      const pc = new OrigRTC(...args);

      pc.addEventListener('track', (event) => {
        if (event.track.kind === 'audio') {
          const stream = event.streams.length > 0 ? event.streams[0] : new MediaStream([event.track]);
          // Use stream.id as the track key (stable across renegotiations)
          const trackKey = stream.id || event.track.id || ('track_' + nextTrackId);

          console.log('[Gneva] WebRTC audio track received: trackKey=' + trackKey +
                       ', track.id=' + event.track.id +
                       ', track.label=' + (event.track.label || 'none') +
                       ', streams=' + event.streams.length +
                       ', track.readyState=' + event.track.readyState +
                       ', track.muted=' + event.track.muted);

          // Create separate processing chain for this track
          createTrackChain(stream, trackKey);

          // Log track state changes for diagnostics
          event.track.addEventListener('mute', () => {
            console.log('[Gneva] Track muted:', trackKey);
          });
          event.track.addEventListener('unmute', () => {
            console.log('[Gneva] Track unmuted:', trackKey);
          });
        } else {
          console.log('[Gneva] Non-audio track received: kind=' + event.track.kind);
        }
      });

      return pc;
    };
    window.RTCPeerConnection.prototype = OrigRTC.prototype;
    Object.getOwnPropertyNames(OrigRTC).forEach(prop => {
      if (prop !== 'prototype' && prop !== 'length' && prop !== 'name') {
        try { window.RTCPeerConnection[prop] = OrigRTC[prop]; } catch(e) {}
      }
    });
  }

  // Handle pre-captured tracks (from the early init_script hook)
  function connectPreCapturedTracks() {
    if (window.__gnevaIncomingAudioTracks && window.__gnevaIncomingAudioTracks.length > 0) {
      console.log('[Gneva] Connecting', window.__gnevaIncomingAudioTracks.length, 'pre-captured audio tracks');
      window.__gnevaIncomingAudioTracks.forEach((stream, idx) => {
        const trackKey = stream.id || ('precaptured_' + idx);
        try {
          createTrackChain(stream, trackKey);
          console.log('[Gneva] Connected pre-captured track:', trackKey);
        } catch (e) {
          console.warn('[Gneva] Pre-captured track connect failed:', e.message);
        }
      });
    }
  }

  // Expose diagnostics globally for debugging from Python
  window.__gnevaAudioDiag = function() {
    const chains = [];
    trackChains.forEach((chain, key) => {
      chains.push({
        trackKey: key,
        trackId: chain.trackId,
        ctxState: chain.ctx.state,
      });
    });
    return {
      wsState: ws ? ws.readyState : -1,
      activeChains: trackChains.size,
      totalTracksEver: nextTrackId - 1,
      chains: chains,
      fallbackActive: !!fallbackCtx,
      connectedElements: connectedElements.size,
    };
  };

  // Start
  connectWS();
  connectPreCapturedTracks();
  scanForMedia();
  setInterval(scanForMedia, 3000);

  // Periodic diagnostic logging
  setInterval(() => {
    const diag = window.__gnevaAudioDiag();
    if (diag.activeChains > 0 || diag.connectedElements > 0) {
      console.log('[Gneva] Audio diag: ' + diag.activeChains + ' track chains, ' +
                   diag.connectedElements + ' DOM elements, WS=' +
                   (ws ? ['CONNECTING','OPEN','CLOSING','CLOSED'][ws.readyState] : 'null'));
    }
  }, 10000);

  console.log('[Gneva] Audio capture initialized (per-track mode), WS:', WS_URL);
})();

File: gneva/bot/audio_capture.py
"""Audio capture server — receives per-track PCM from injected JS via WebSocket, writes WAV."""

import asyncio
import struct
import wave
import logging
import os
from pathlib import Path
from typing import Callable

logger = logging.getLogger(__name__)

AUDIO_JS_PATH = Path(__file__).parent / "audio_capture.js"

# Track ID 0 is the fallback mixed-audio path (DOM element capture)
FALLBACK_TRACK_ID = 0


class AudioCapture:
    """WebSocket server that receives per-track PCM16 audio chunks and writes them to WAV files.

    Each incoming WebSocket message has a 4-byte big-endian track ID header followed
    by PCM16 audio data. Tracks are kept separate so each speaker can be transcribed
    independently.
    """

    def __init__(self, output_path: str, sample_rate: int = 16000):
        self.output_path = output_path
        self.sample_rate = sample_rate
        self.channels = 1
        self._server = None
        self._port = 0
        # Per-track audio storage: track_id -> list[bytes]
        self._track_chunks: dict[int, list[bytes]] = {}
        # Combined chunks for WAV output (all tracks mixed)
        self._chunks: list[bytes] = []
        self._total_bytes = 0
        self._running = False
        # Callback for real-time STT — receives (pcm_bytes, track_id)
        self._on_audio_chunk: Callable[[bytes, int], None] | None = None
        # Legacy callback without track_id (for backward compat)
        self._on_audio_chunk_legacy: Callable[[bytes], None] | None = None

    @property
    def port(self) -> int:
        return self._port

    def get_ws_url(self) -> str:
        return f"ws://127.0.0.1:{self._port}"

    def get_inject_js(self) -> str:
        """Return the audio capture JS with the WebSocket URL injected."""
        js = AUDIO_JS_PATH.read_text(encoding="utf-8")
        return js.replace("__WS_URL__", self.get_ws_url())

    async def start(self):
        """Start the WebSocket server on a random available port."""
        import websockets

        self._running = True
        self._server = await websockets.serve(
            self._handle_connection,
            "127.0.0.1",
            0,  # random port
        )
        # Get the actual port
        for sock in self._server.sockets:
            addr = sock.getsockname()
            self._port = addr[1]
            break

        logger.info(f"Audio capture WebSocket started on port {self._port}")

    def set_audio_chunk_callback(self, callback: Callable[[bytes, int], None]):
        """Set a callback to receive raw PCM chunks with track ID (for per-speaker STT).

        Callback signature: callback(pcm_bytes: bytes, track_id: int)
        """
        self._on_audio_chunk = callback

    def set_legacy_audio_chunk_callback(self, callback: Callable[[bytes], None]):
        """Set a legacy callback that receives raw PCM chunks without track ID."""
        self._on_audio_chunk_legacy = callback

    async def _handle_connection(self, websocket):
        """Handle incoming WebSocket connection from injected JS.

        Each binary message has format: [4-byte uint32 BE track_id][PCM16 audio data]
        """
        logger.info("Audio capture: browser connected")
        msg_count = 0
        tracks_seen = set()

        try:
            async for message in websocket:
                if not isinstance(message, bytes) or not self._running:
                    continue

                msg_count += 1

                # Parse track ID header (4 bytes big-endian uint32)
                if len(message) <= 4:
                    continue

                track_id = struct.unpack('>I', message[:4])[0]
                pcm_data = message[4:]

                # Log first message from each track
                if track_id not in tracks_seen:
                    tracks_seen.add(track_id)
                    logger.info(
                        f"Audio capture: first data from track {track_id} "
                        f"({len(pcm_data)} bytes, total tracks: {len(tracks_seen)})"
                    )

                # Store per-track
                if track_id not in self._track_chunks:
                    self._track_chunks[track_id] = []
                self._track_chunks[track_id].append(pcm_data)

                # Also store combined for WAV output
                self._chunks.append(pcm_data)
                self._total_bytes += len(pcm_data)

                # Forward to real-time STT with track ID
                if self._on_audio_chunk:
                    try:
                        self._on_audio_chunk(pcm_data, track_id)
                    except Exception:
                        pass
                elif self._on_audio_chunk_legacy:
                    try:
                        self._on_audio_chunk_legacy(pcm_data)
                    except Exception:
                        pass

                # Periodic logging
                if msg_count % 500 == 0:
                    logger.debug(
                        f"Audio capture: {msg_count} messages, "
                        f"{self._total_bytes / 1024:.0f} KB, "
                        f"{len(tracks_seen)} tracks"
                    )

        except Exception as e:
            logger.debug(f"Audio WS connection closed: {e}")

        logger.info(
            f"Audio capture: connection ended — "
            f"{msg_count} messages, {len(tracks_seen)} tracks, "
            f"{self._total_bytes / 1024:.0f} KB total"
        )

    def get_track_ids(self) -> list[int]:
        """Return list of track IDs that have received audio."""
        return list(self._track_chunks.keys())

    async def stop(self) -> str:
        """Stop capture, write WAV file, return the file path."""
        self._running = False

        if self._server:
            self._server.close()
            await self._server.wait_closed()

        # Write combined WAV (all tracks mixed together)
        os.makedirs(os.path.dirname(self.output_path), exist_ok=True)

        if not self._chunks:
            logger.warning("No audio data captured")
            return self.output_path

        with wave.open(self.output_path, "wb") as wf:
            wf.setnchannels(self.channels)
            wf.setsampwidth(2)  # 16-bit = 2 bytes
            wf.setframerate(self.sample_rate)
            for chunk in self._chunks:
                wf.writeframes(chunk)

        duration_sec = self._total_bytes / (self.sample_rate * 2 * self.channels)
        track_summary = ", ".join(
            f"track{tid}={len(chunks)}chunks"
            for tid, chunks in self._track_chunks.items()
        )
        logger.info(
            f"Audio saved: {self.output_path} "
            f"({duration_sec:.1f}s, {self._total_bytes / 1024:.0f} KB, "
            f"tracks: {track_summary})"
        )
        return self.output_path

    @property
    def duration_sec(self) -> float:
        if not self._total_bytes:
            return 0.0
        return self._total_bytes / (self.sample_rate * 2 * self.channels)

    @property
    def has_audio(self) -> bool:
        return self._total_bytes > 0
